Zend/zend_gc.c

```diff
@@ -261,28 +261,45 @@ struct _gc_stack {
 	size_t    _top = 0;
 
 #define GC_STACK_PUSH(ref) \
-	do { \
-		if (UNEXPECTED(_top == GC_STACK_SEGMENT_SIZE)) { \
-			if (UNEXPECTED(!_stack->next)) { \
-				gc_stack *segment = emalloc(sizeof(gc_stack)); \
-				segment->prev = _stack; \
-				segment->next = NULL; \
-				_stack->next = segment; \
-			} \
-			_stack = _stack->next; \
-			_top = 0; \
-		} \
-		_stack->data[_top++] = (ref); \
-	} while (0)
+	gc_stack_push(&_stack, &_top, ref);
 
 #define GC_STACK_POP() \
-	(UNEXPECTED(_top == 0) \
-		? ((!_stack->prev) \
-			? NULL \
-			: ( _stack = _stack->prev, \
-				_top = GC_STACK_SEGMENT_SIZE - 1, \
-				_stack->data[_top])) \
-		: _stack->data[--_top])
+	gc_stack_pop(&_stack, &_top)
+
+static zend_never_inline gc_stack* gc_stack_next(gc_stack *stack)
```

Comment:
why is this `zend_never_inline`? It's used in only one place so it could have been manually inlined there so I imagine it could be inlined automatically also.

Reply:
It's used in only one place, but that place is an always-inline function. We want to inline the fast-path of stack pushing to each use, but we don't want to inline the rarely called reallocation code.

Reply:
right. It's not inlined on purpose, to reduce code size and improve register allocation on fast path.